src/renderers/dom/fiber/ReactDOMFiberComponent.js

```diff
@@ -227,6 +238,9 @@ function setInitialDOMProperties(
       // Noop
     } else if (registrationNameModules.hasOwnProperty(propKey)) {
       if (nextProp) {
+        if (__DEV__ && typeof nextProp !== 'function') {
```

Comment:
You can move this assertion into the `warnForInvalidEventListener` function. E.g. as the first argument to `warning()`.